### PR TITLE
Add basic file drag and drop support

### PR DESCRIPTION
**Pull request by ShadowMitia**
Adds basic drag and drop support to bevy.

It simply exposes the relevant events already found in winit. 
This has been tested on Linux, and because it simply wraps around the winit events, it should work on other systems just fine (hopefully!).

I still want to try wasm/web before merging. (Or someone else, I haven't played with wasm in rust or bevy much yet!)

Let me know if there's anything missing!

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -165,6 +165,10 @@ path = "examples/app/return_after_run.rs"
 name = "thread_pool_resources"
 path = "examples/app/thread_pool_resources.rs"
 
+[[example]]
+name = "draganddrop"
```
Comment (smokku) [inline Cargo.toml:169]: Can we keep examples sorted alphabetically?

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -350,6 +351,35 @@ pub fn winit_runner(mut app: App) {
                         ),
                     }
                 }
+                WindowEvent::DroppedFile(path_buf) => {
+                    let mut cursor_left_events =
```
Comment (smokku) [inline crates/bevy_winit/src/lib.rs:355]: It should be called `dropped_file_events`, right?

File: crates/bevy_winit/src/lib.rs [hunk 3]
```diff
@@ -370,6 +400,7 @@ pub fn winit_runner(mut app: App) {
                 );
                 app.update();
             }
+
```
Comment (smokku) [inline crates/bevy_winit/src/lib.rs:403]: Other matches are not separated.

**Comment by DJMcNab** [comment]
Could you clarify in the title that this is file drag and drop support?

**Comment by ShadowMitia** [comment]
@DJMcNab Will do. I'll update the commit message as well.

File: Cargo.toml (hunk 1 above)
Comment (ShadowMitia) [reply Cargo.toml:169]: I never realised they were in alphabetical order...

File: crates/bevy_winit/src/lib.rs (hunk 2 above)
Comment (ShadowMitia) [reply crates/bevy_winit/src/lib.rs:355]: Indeed. Also fixed this in the other functions.

File: crates/bevy_winit/src/lib.rs (hunk 3 above)
Comment (ShadowMitia) [reply crates/bevy_winit/src/lib.rs:403]: I'm not sure what you're saying here sorry, could you explain please?

**Comment by ShadowMitia** [comment]
I've tested on wasm, and this doesn't work unfortunently. But it looks like winit doesn't have it either https://github.com/rust-windowing/winit/blob/master/FEATURES.md#input-handling-1.

File: Cargo.toml [hunk 4]
```diff
@@ -133,6 +133,11 @@ path = "examples/3d/z_sort_debug.rs"
 name = "custom_loop"
 path = "examples/app/custom_loop.rs"
 
+[[example]]
+name = "draganddrop"
```
Comment (cart) [inline Cargo.toml:137]: Can we rename the example / file to "drag_and_drop"? I like to use snake_case for all file names.

File: crates/bevy_winit/src/lib.rs [hunk 5]
```diff
@@ -350,6 +351,35 @@ pub fn winit_runner(mut app: App) {
...
+                    let winit_windows = app.resources.get_mut::<WinitWindows>().unwrap();
+                    let window_id = winit_windows.get_window_id(winit_window_id).unwrap();
+                    dropped_file_events.send(DroppedFile {
+                        id: window_id,
+                        path_buf,
+                    });
+                }
+                WindowEvent::HoveredFile(path_buf) => {
+                    let mut hovered_file_events =
+                        app.resources.get_mut::<Events<HoveredFile>>().unwrap();
```
Comment (cart) [inline crates/bevy_winit/src/lib.rs:366]: I think we should probably combine all drag and drop events into a single event. It seems like order of operations matters here (ex: when did the cancel happen).

(not relevant to this pr, but there are a couple of other events here that we should revisit when it comes to combining/ordering, such as CursorLeft/CursorEntered)

**Comment by ShadowMitia** [comment]
I took a stab at creating a simple plugin and resource to handle dropped files (and added it to `DefaultPlugins).
Is this the idea you had in mind @cart? I'm not compeltely satisfied with the interface, but then I'm still kinda new at Rust.

The plugin gives you a `DragAndDrop` resource, which you can iterate using a `drain` method.
Each file is a tuple of `(PathBuf, and WindowId)`. If a path is a directory, the user is left to manage what to do.

I can't see the point of allowing direct access to `HoveredFiles` for now so I didn't create an interface for it, but maybe I'm wrong.
There is an `is_hovered` method if you want to do something special during hovering. (I'm thinking changing the cursor icon mainly or have some kind of feedback).

**Comment by ShadowMitia** [comment]
For reference, I'm also attempting to add file drag and drop into winit https://github.com/rust-windowing/winit/pull/1808

**Comment by cart** [comment]
I was more talking about combining the drag and drop events into a single event stream to preserve ordering (see the changes I just pushed).

I think the drag and drop plugin introduces a bit too much complexity for what it adds, so I reverted that in favor of the consolidated event.

**Comment by cart** [comment]
Just rebased on master / resolved conflicts. I think this is good to go!

**Comment by ShadowMitia** [comment]
I fixed the path in the README and changed the system name in the example to something a bit clearer.
Otheriwse I think it's all good (untill I get wasm working anyway, if it happens :grin: )

**Pull request merged**